File: scripts/Python/lab_06_biomech_data.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Set the working directory (if needed)
import os
os.chdir("C:/Users/kelop/OneDrive/Documents/GitHub/PhysTher5110/")
os.listdir()

# # Read the raw data
# raw_dat = pd.read_csv("./data/gait_example_data/DDH25_0Run01.csv", sep=",",
# header=None)
# 
# # Reshape and label the FORCE data
# force_dat = raw_dat.iloc[5:9794, :]
# force_labs = raw_dat.iloc[2:4, :].T
# force_labs.columns = ["level1", "level2"]
# force_labs['level1'].fillna(method='ffill', inplace=True)
# force_labs['level3'] = force_labs['level1'] + "_" + force_labs['level2']
# force_dat.columns = force_labs['level3']
# force_dat.to_csv("./data/gait_example_data/force_data.csv", index=False)
# 
# # Reshape and label the MOTION data
# motion_dat = raw_dat.iloc[3:, :]
# motion_labs = raw_dat.iloc[:2, :].T
# motion_labs.columns = ["level1", "level2"]
# motion_labs['level1'].fillna(method='ffill', inplace=True)
# motion_labs['level3'] = motion_labs['level1'] + "_" + motion_labs['level2']
# motion_dat.columns = motion_labs['level3']
# motion_dat.to_csv("./data/gait_example_data/motion_data.csv", index=False)
# 

# Select and rename variables
force_dat = pd.read_csv("./data/gait_example_data/force_data.csv", sep=",")

force_dat = force_dat[['Treadmill Left - Force_Fx', 'Treadmill Left - Force_Fy', 
                        'Treadmill Left - CoP_Cx', 'Treadmill Left - CoP_Cy']]
force_dat.columns = ['force_x', 'force_y', 'cop_x', 'cop_y']
force_dat['sample'] = force_dat.index


# Plot FORCE data
plt.figure(figsize=(10, 6))
plt.plot(force_dat['force_y'], linestyle='-', label='Force Y')
plt.plot(force_dat['force_x'], linestyle='-', label='Force X')
plt.plot(force_dat['cop_y'], linestyle='-', label='CoP Y')
plt.plot(force_dat['cop_x'], linestyle='-', label='CoP X')
plt.xlabel('Sample')
plt.ylabel('Value')
plt.legend(loc="best")
plt.show()

# Reshape and rename MOTION data
motion_dat = pd.read_csv("./data/gait_example_data/motion_data.csv", sep=",")

motion_dat = motion_dat[['DDH25:RICAL_X', 'DDH25:RICAL_Y', 'DDH25:LICAL_X', 'DDH25:LICAL_Y']]
motion_dat.columns = ['right_heel_x', 'right_heel_y', 'left_heel_x', 'left_heel_y']
motion_dat['sample'] = motion_dat.index



# Plot MOTION data
plt.figure(figsize=(10, 6))
plt.plot(motion_dat['right_heel_x'], linestyle='-', label='Right Heel X')
plt.plot(motion_dat['right_heel_y'], linestyle='-', label='Right Heel Y')
plt.plot(motion_dat['left_heel_x'], linestyle='-', label='Left Heel X')
plt.plot(motion_dat['left_heel_y'], linestyle='-', label='Left Heel Y')
plt.xlabel('Sample')
plt.ylabel('Value')
plt.legend(loc='best')
plt.show()

# Downsample the FORCE data
force_dat_ds = force_dat.iloc[::10, :]

# Check the number of rows in both datasets
print(len(force_dat_ds))
print(len(motion_dat))

# Merge the datasets
merged = pd.concat([motion_dat, force_dat_ds.iloc[:, :-1]], axis=1)
merged = merged.apply(pd.to_numeric, errors='ignore')

# Basic plots and summary statistics
print(merged['sample'].describe())
print(merged.info())

# Plotting heel positions and force
plt.figure(figsize=(10, 6))
plt.plot(merged['sample'], merged['right_heel_y'], label='Right Heel Y', linestyle='-')
plt.plot(merged['sample'], merged['left_heel_y'], label='Left Heel Y', linestyle='-')
plt.xlabel('Sample')
plt.ylabel('Heel Position')
plt.legend()
plt.show()

plt.figure(figsize=(10, 6))
plt.scatter(merged['left_heel_y'], merged['right_heel_y'], label='Left vs. Right Heel')
plt.xlabel('Left Heel Y')
plt.ylabel('Right Heel Y')
plt.legend()
plt.show()

plt.figure(figsize=(10, 6))
plt.plot(merged['right_heel_y'], merged['force_y'], linestyle='-', label='Force Y vs. Right Heel Y')
plt.xlabel('Right Heel Y')
plt.ylabel('Force Y')
plt.legend()
plt.show()

File: scripts/MATLAB/Lab_06_biomech_data.m
% % Set the working directory (modify the path accordingly)
% cd('~/GitHub/PhysTher5110/');
% 
% % Read the raw data
% RAW_DAT = csvread('./data/gait_example_data/DDH25_0Run01.csv', 5, 0);
% 
% % Reshape and label the FORCE data
% FORCE_DAT = RAW_DAT(:, 1:9790);
% FORCE_LABS = RAW_DAT(3:4, :);
% FORCE_LABS = FORCE_LABS.';
% FORCE_LABS = array2table(FORCE_LABS, 'VariableNames', {'level1', 'level2'});
% FORCE_LABS.level1 = fillmissing(FORCE_LABS.level1, 'previous');
% FORCE_LABS.level3 = strcat(FORCE_LABS.level1, '_', FORCE_LABS.level2);
% colnames = FORCE_LABS.level3;
% FORCE_DAT = array2table(FORCE_DAT, 'VariableNames', colnames);
% writetable(FORCE_DAT, './data/gait_example_data/force_data.csv');
% 
% % Reshape and label the MOTION data
% RAW_DAT = csvread('./data/gait_example_data/DDH25_0Run01.csv', 9797, 0);
% MOTION_DAT = RAW_DAT(:, 3:end);
% MOTION_LABS = RAW_DAT(1:2, :);
% MOTION_LABS = MOTION_LABS.';
% MOTION_LABS = array2table(MOTION_LABS, 'VariableNames', {'level1', 'level2'});
% MOTION_LABS.level1 = fillmissing(MOTION_LABS.level1, 'previous');
% MOTION_LABS.level3 = strcat(MOTION_LABS.level1, '_', MOTION_LABS.level2);
% colnames = MOTION_LABS.level3;
% MOTION_DAT = array2table(MOTION_DAT, 'VariableNames', colnames);
% writetable(MOTION_DAT, './data/gait_example_data/motion_data.csv');

FORCE_DAT = readtable('force_data.csv', 'VariableNamingRule','preserve');
MOTION_DAT = readtable('motion_data.csv', 'VariableNamingRule','preserve');

% Select and rename variables
FORCE_DAT = FORCE_DAT(:, {'Treadmill Left - Force_Fx', 'Treadmill Left - Force_Fy',...
    'Treadmill Left - CoP_Cx', 'Treadmill Left - CoP_Cy'});
FORCE_DAT.Properties.VariableNames = {'force_x', 'force_y', 'cop_x', 'cop_y'};

MOTION_DAT = MOTION_DAT(:, {'DDH25:RICAL_X', 'DDH25:RICAL_Y', 'DDH25:LICAL_X', 'DDH25:LICAL_Y'});
MOTION_DAT.Properties.VariableNames = {'right_heel_x', 'right_heel_y', 'left_heel_x', 'left_heel_y'};

% Plot the data (you can customize the plots as needed)
figure;
plot(FORCE_DAT.force_y);
title('FORCE_DAT: force_y');

figure;
plot(FORCE_DAT.force_x);
title('FORCE_DAT: force_x');

figure;
plot(FORCE_DAT.cop_y);
title('FORCE_DAT: cop_y');

figure;
plot(FORCE_DAT.cop_x);
title('FORCE_DAT: cop_x');

figure;
plot(MOTION_DAT.right_heel_x);
title('MOTION_DAT: right_heel_x');

figure;
plot(MOTION_DAT.right_heel_y);
title('MOTION_DAT: right_heel_y');

figure;
plot(MOTION_DAT.left_heel_x);
title('MOTION_DAT: left_heel_x');

figure;
plot(MOTION_DAT.left_heel_y);
title('MOTION_DAT: left_heel_y');

% Down sample the FORCE data
FORCE_DAT_DS = FORCE_DAT(1:10:end, :);

% Merge the datasets
MERGED = [MOTION_DAT, FORCE_DAT_DS];
MERGED.sample = (1:size(MERGED, 1))';

% Basic plots
% figure;
% hist(MERGED.sample);
% title('Sample Histogram');

figure;
scatter(MERGED.right_heel_y, MERGED.left_heel_y);
title('Right vs. Left Heel Position');

figure;
scatter(MERGED.force_y, MERGED.right_heel_y);
title('Force vs. Right Heel Position');
